rfcs/0000_intelligent_resampling.md

```diff
@@ -0,0 +1,368 @@
+
+This scenario is already handled by the existing `hypotheses_displacer.py`, which computes evidence based on feature mismatches. No changes are needed for Case 2, except to handle null observations from Case 1b.
+
+### Implications for FeatureChangeSM
+
+**Current limitation**: As mentioned earlier, `FeatureChangeSM` currently filters out off-object observations, preventing the LM from receiving them. This filtering happens in the `check_feature_change()` method, which returns `False` for off-object observations. To enable prediction error-based hypothesis elimination for null observations (Case 1b), we need to modify how `FeatureChangeSM` handles these observations.
+
+The following changes would be made to `sensor_module.py`:
+
+**1. Handling `check_feature_change()` in `sensor_modules.py`**
+```python
+# the if statement below needs update
+if not observed_features.get_on_object():
+    # Even for the surface-agent sensor, do not return a feature for LM
+    # processing that is not on the object
+    logger.debug(f"No new point because not on object")
+    return False
+```
+
+Expected behavior with the proposed changes:
+- **On-object to on-object**: Feature changes will continue to work as currently implemented
+- **On-object to off-object (and vice versa)**: Will be treated as significant changes; the learning module will receive "null" features when transitioning from on-object to off-object
+- **Off-object to off-object**: Same as on-object to on-object - there should be no feature change (may need to handle different in NaN values)
+
+**3. How `on_object` is used in SM**
+
+- The `on_object` is set to `float(semantic_id > 0)` where `semantic_id` indicates if an object is detected at the sensor's center point
+- The `use_state` flag combines `on_object` status AND signal validity
+```python
+use_state=bool(morphological_features["on_object"]) and not invalid_signals
+```
+A `False` value currently prevents null observations from being processed by LM. With the proposed RFC, this should only be set to False when we go from off-object to off-object transition.
+
+### When we're learning an object, we don't have a complete graph model. How do we deal with this?
+
+During object learning, prediction errors present a complex decision point: they may indicate (1) incorrect hypotheses that should be eliminated, (2) an incomplete model that needs updating, or (3) both. These interpretations are not mutually exclusive - a large prediction error might mean we should eliminate current hypotheses AND learn or update our models.
+
+This challenge becomes particularly complex when learning and inference are interleaved, or when the distinction between "pure learning" and "pure inference" modes is not clearly defined. In pure inference mode, we may use prediction errors to eliminate hypotheses assuming the object model is considered complete. In pure learning mode, prediction errors may indicate model needs updating since the model is incomplete.
```

Comment:
```suggestion
This challenge becomes particularly complex when learning and inference are interleaved. In evaluation mode (no model updates), we may use prediction errors to eliminate hypotheses assuming the object model is considered complete. In learning mode, prediction errors may indicate model needs updating since the model is incomplete.
```

Reply:
I'm not sure this paragraph is useful. It seems like the same issues would happen in learning and evaluation mode. The only difference may be if we assume that we only evaluate on complete models but I'm not sure this assumption is useful.